### ISSUE TITLE
Add tabindex to main element landmark

### ISSUE THREAD

**Issue by schalkneethling** [mdn/yari#2755]
In the header of MDN Web Docs we provide context-sensitive skips links to the main content, language selector(Only on the homepage. There are two bugs here which I will log separately), and search. For non-natively focusable elements, we need to add `tabindex="-1"` to make them focusable.

The search link links to the search input which is natively focusable so we are all good here. When we address the link to the language selector we need to ensure it links to the `select` element. The `main` element is not natively focusable and so it needs the addition of `tabindex="-1"`

Why? Well,

> Some browsers, such as Safari, require the link destination to either be natively focusable (such as a link, button, or form element), or have a tabindex value to function correctly. If the link is not focusable and does not have a tabindex value, the viewport will scroll down to the appropriate place, but when the user presses the tab key, the focus will jump back up to the next link after the "skip navigation" link, when it should jump to the next link after the destination.

See additional context:
https://github.com/mdn/yari/issues/2755#issuecomment-770586949

**Comment by Ryuno-Ki** [comment]
Pardon me, but [tabindex](https://developer.mozilla.org/en-us/docs/Web/HTML/Global_attributes/tabindex) with value -1 means

> A negative value (usually `tabindex="-1"`) means that the element is not reachable via sequential keyboard navigation, but could be focused with Javascript or visually by clicking with the mouse. It's mostly useful to create accessible widgets with JavaScript. 

Perhaps worth to use `tabindex="0"` instead?

**Comment by mattiapontonio** [comment]
https://github.com/mdn/yari/blob/0f36b392c8f755144722c0ee9661cd65138f3fe2/client/src/ui/molecules/a11y-nav/index.tsx

This? But sorry what is it for?

**Comment by Ryuno-Ki** [comment]
Because you are dealing with React here.
React is based on components.
Those components render HTML (or other React components).

There is a special syntax in React called JSX (= JavaScript eXtended). With TypeScript support it becomes TSX (= TypeScript eXtended).

**Comment by mattiapontonio** [comment]
What does React have to do with it? I was just saying that TalkBack can already jump from header to header or paragraph to paragraph and maybe there is no need to maintain functionality but simply provide visible and accessible HTML.

**Comment by Ryuno-Ki** [comment]
Please specify your question clearer then :-)

> This? But sorry what is it for?

I read this as:

> Is this the file in need of change? Why is it a TSX?

resp.

> Why an a11n-nav and not a normal accessible nav just?

as

> Cannot you use a plain old HTML? With a `<nav>`.

The answer in a React project is: no.

---

> What does React have to do with it?

Yari is built with React.

> I was just saying that TalkBack can already jump from header to header or paragraph to paragraph […]

That's right. But you're dealing with a link target here. That needs to be focussable, otherwise you'll loose focus (as described in the initial post). `tabindex="0"` is doing just that.

> […] maybe there is no need to maintain functionality but simply provide visible and accessible HTML

Not in the setup chosen by the MDN team.
You _could_ use part of the website with React only (and have the rest be generated statically). But it was an architectural decision to go another route (and do everything with React).

> However yes I confirm that both React and Typescript could be removed but now it is practically impossible, the best that can be done is boycott them.

Unrelated to this issue (better to open a discussion if you want to go into it).

**Comment by mattiapontonio** [comment]
No, it was all simpler. 
The question was: Is this the component? But don't we first eliminate it? 

---

Regarding the architectural discussions ... 👋

**Comment by Ryuno-Ki** [comment]
Or you could become broccoli and ban it: https://www.nytimes.com/1990/03/23/us/i-m-president-so-no-more-broccoli.html

Can we focus on the issue, please?

**Comment by mattiapontonio** [comment]
Yes You was right regarding tabindex
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex
👍

**Comment by Ryuno-Ki** [comment]
As far as I can tell, `#content` is defined in

https://github.com/mdn/yari/blob/21d7ecc26d2466713481ddc2f5b27f3023cdb4a5/client/src/ui/atoms/page-content/index.tsx#L12-L18

and `#main-q` in

https://github.com/mdn/yari/blob/12892a53a452d70f285e0f2383fe586aecfbf837/client/src/ui/molecules/search/index.tsx#L54-L62

but also in

https://github.com/mdn/yari/blob/35a4654f4edde9fb13c81cd3186930dd1f577de1/client/src/search.tsx#L277-L301

(so perhaps to submit this as an additional PR?)

**Comment by mattiapontonio** [comment]
I understood what we were talking about
- https://github.com/mdn/yari/search?q=id%3Dskip-main&type=
- https://github.com/mdn/yari/search?q=id%3Dskip-search&type=

**Comment by Ryuno-Ki** [comment]
Not quite.
Those IDs are defined in

https://github.com/mdn/yari/blob/0f36b392c8f755144722c0ee9661cd65138f3fe2/client/src/ui/molecules/a11y-nav/index.tsx#L28-L49

In order to keep focus for Assistive Technology, the jump target needs to be focusable.
Some elements (like `<a>`, `<input />` or `<button>`) do so natively. There is nothing to do here once the _skip link_ was activated.

However, a `<section>` cannot receive the focus per se. Once it has `tabindex="0"` that changes.

**Comment by mattiapontonio** [comment]
Ok now I understand the problem. However, I wonder if these links should be visible for some other accessibility issue such as not being seen.

**Comment by Ryuno-Ki** [comment]
One thing at a time. This issue here is about adding `tabindex`.
Feel free to open another one to talk about visibility.

**Comment by mattiapontonio** [comment]
Also, what is the point of giving focus to an object that is neither a command nor a paragraph to read?

---

No I don't open another issue, that's okay, I would like to go towards another path at the moment.

**Comment by Ryuno-Ki** [comment]
The point is

> If the link is not focusable and does not have a tabindex value, the viewport will scroll down to the appropriate place, but when the user presses the tab key, the focus will jump back up to the next link after the "skip navigation" link, when it should jump to the next link after the destination.

That means, you would end up in the header once you <kbd>Tab</kbd> after having used the skip link. An unpleasant user experience.

**Comment by mattiapontonio** [comment]
Couldn't it be related to the fact that it's invisible?
Does the same thing happen in the Jump to section that is visible?

**Comment by Ryuno-Ki** [comment]
Try it out and report back :-)

**Comment by mattiapontonio** [comment]
I can't activate Skip to main content from mobile because I can't see it 👁️❌, I can only hear it 👂

**Comment by Ryuno-Ki** [comment]
Guess what: It has no `tabindex="0"` ;-)

**Comment by mattiapontonio** [comment]
> Guess what: It has no `tabindex="0"` ;-)

Ok but then we basically have to add the tabindex wherever we have an ID, I'm not very convinced.

**Comment by Ryuno-Ki** [comment]
Nobody forces you to submit a PR to this issue.

**Comment by mattiapontonio** [comment]
FTR
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/main#accessibility_concerns

**Comment by Ryuno-Ki** [comment]
I'm assuming you're referring to this line?

> Adding an `id` attribute to the `<main>` element lets it be a target of a skip navigation link.

Hm, the referenced WebAIM article is from 2013.

**Comment by mattiapontonio** [comment]
But in the end what do we want here? That the main receives the focus in the correct order in which it is and behaves like the others even though it is not a normal element that receives the focus?

**Comment by Ryuno-Ki** [comment]
> I can't activate **Skip to main content** from mobile because I can't see it eyex and touch It point_leftx I can only hear it earheavy_check_mark
> 
> On the **Jump to section** a strange thing happened to me, as if the focus dart went on the section but the swipe right point_up_2arrow_right returned to the beginning arrow_up of the document.

We want to improve this.

**Comment by mattiapontonio** [comment]
First I should see the mobile links in order to activate them, then from there I could test.
But can they be seen on the desktop? Only on the focus?

**Comment by Ryuno-Ki** [comment]
Would you mind to open an issue to make skip links visible?
We can declare this ticket as being blocked by the new issue then.
Thanks.

**Comment by mattiapontonio** [comment]
We could also try to solve this issue incorrectly and close it and wait.

**Comment by Ryuno-Ki** [comment]
Why? The issue still stands.
It's a oneline change.

**Comment by mattiapontonio** [comment]
`tabindex=0` on `main`, that's all.

**Comment by schalkneethling** [comment]
> Pardon me, but [tabindex](https://developer.mozilla.org/en-us/docs/Web/HTML/Global_attributes/tabindex) with value -1 means

I appreciate the confusion here but, `tabindex=-1` is indeed what we want.

The aim here is to make an element that is not natively focusable, focusable but, we do not want to introduce it to the tab order. Setting `tabindex` to `-1` will achieve this.

If we set this to `tabindex=0` we are not only making it focusable but also adding it to the tab order. But there is a catch 😄  as described here:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex

> tabindex="0" means that the element should be focusable in sequential keyboard navigation, after any positive tabindex values and its order is defined by the document's source order.

That means if we add `tabindex=0` to the `main` element and a user tabs through all focusable elements(or reaches the last focusable element by different means), the next tab press will not cycle back to the first element but, will first move to the `main` element which is not what we want.

Using `tabindex` with a positive value is, as you know, generally a bad idea and should pretty much never be used.

Thanks for highlighting this @Ryuno-Ki, I will add a link to my further explanation here to the original issue description above.

**Comment by schalkneethling** [comment]
> * Why an a11n-nav and not a normal accessible nav just?

This was mostly a design decision back when the original redesign was done. We could, and still can opt to just make these elements always visible and not just on tab focus, but that would be a different discussion 😃

**Comment by schalkneethling** [comment]
> (so perhaps to submit this as an additional PR?)

No need for this as an `input` element is natively focusable.

**Comment by schalkneethling** [comment]
> I was just saying that TalkBack can already jump from header to header or paragraph to paragraph and maybe there is no need to maintain functionality but simply provide visible and accessible HTML.

Yes on both points with regards to "visible and accessible HTML" but:

1. The visible part is sometimes a design decision and so we need to employ different techniques to ensure things are still accessible
2. TalkBack is on Android, my quoted comment mentions Safari so, this is more to address an issue with Safari and VoiceOver.

**Comment by Ryuno-Ki** [comment]
From what I can confirm: The issue happens the same on Firefox for Android with TalkBack.
The quote in your first article mentions „Some browsers, such as Safari” - so it's not limited to it.

**Comment by schalkneethling** [comment]
> From what I can confirm: The issue happens the same on Firefox for Android with TalkBack.
> The quote in your first article mentions „Some browsers, such as Safari” - so it's not limited to it.

For sure, but the solution proposed here will cover all browsers and platforms. Thank you for confirming.

**Comment by mattiapontonio** [comment]
> we need to employ different techniques

I listened 👂 to the speech with interest. I am curiously ignorant of seeing 👁️ the final result.

**Comment by schalkneethling** [comment]
Looks like this is not needed for us. This applies to pretty old version of Safari and other browsers. I reckon we are fine as the behavior in our supported browsers are as expected without setting `tabindex`
